src/Compilers/Core/Portable/Compilation.EmitStreamProvider.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// The compilation object is an immutable representation of a single invocation of the
+    /// compiler. Although immutable, a compilation is also on-demand, and will realize and cache
+    /// data as necessary. A compilation can produce a new compilation from existing compilation
+    /// with the application of small deltas. In many cases, it is more efficient than creating a
+    /// new compilation from scratch, as the new compilation can reuse information from the old
+    /// compilation.
+    /// </summary>
+    public abstract partial class Compilation
+    {
+        /// <summary>
+        /// Abstraction that allows the caller to delay the creation of the <see cref="Stream"/> values 
+        /// until they are actually needed.
+        /// </summary>
+        internal abstract class EmitStreamProvider
```

Comment:
@AlekseyTs 

The original design had the compiler close the `Stream` but that turned out to be untennable as the majority of callers expect to maintain ownership of their `Stream` objects.  When the compiler closed / disposed it resources expected to be available after the compilation were not available.  In fact the only caller for which the compiler closing the `Stream` was advantageous was this one.  Hence the current design. 

The motivation here for the abstraction was not to reduce parameters but instead make it easier to cache the created values.  From the perspective of the compiler this interface is no more usable than say `Func<Stream>` for each parameter.  But from the caller point of view I found it make the interface much easier to call if I had an object on which to store items like the created `Stream`.

I prefer to keep the `HasPdbStream` property because there are different semantics for the two `Stream` creation methods.

- PE Stream can never be `null`.  If `null` is returned it's a violation of the API. 
` PDB Stream can be `null`.  A `null` return without the `HasPdbStream` property makes the return ambiguous.  Was it `null` because there was no `Stream` or was it `null` because of an underlying error?